llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -3141,6 +3141,72 @@ static bool isNonEqualSelect(const Value *V1, const Value *V2, unsigned Depth,
          isKnownNonEqual(SI1->getFalseValue(), V2, Depth + 1, Q);
 }
 
+static bool isNonEqualPointersWithRecursiveGEP(const Value *A, const Value *B,
+                                               unsigned Depth,
+                                               const SimplifyQuery &Q) {
+  // Where A is a recursive GEP for an incoming value of PHI indicating a
+  // loop. B can be a ptr/GEP.
+  // If the PHI has 2 incoming values, one of it being the recursive GEP A
+  // and other a ptr at same base and at an same/higher offset than B we are
+  // only incrementing the pointer further in loop if offset of recursive GEP is
+  // greater than 0.
+  if (!A->getType()->isPointerTy() || !B->getType()->isPointerTy())
+    return false;
+
+  auto *GEPA = dyn_cast<GEPOperator>(A);
+  // Make sure GEPA pointer operand is a PHI.
+  if (!GEPA || !isa<PHINode>(GEPA->getPointerOperand()))
+    return false;
+
+  // Handle 2 incoming PHI values with one being a recursive GEP.
+  auto *PN = dyn_cast<PHINode>(GEPA->getPointerOperand());
+  if (PN->getNumIncomingValues() == 2) {
+    // Recursive GEP in second incoming value. Always keep Recursive GEP as
+    // FirstInst
+    Value *FirstInst = nullptr, *SecondInst = nullptr;
+    for (unsigned i = 0; i != 2; ++i) {
+      // Check if A is a Recursive GEP in one of the incoming values.
+      if (PN->getIncomingValue(i) == A && GEPA->getNumIndices() == 1 &&
+          isa<Constant>(GEPA->idx_begin())) {
+        FirstInst = PN->getIncomingValue(i);
+        SecondInst = PN->getIncomingValue(1 - i);
+        continue;
+      }
+    }
+    if (FirstInst && SecondInst) {
+      // Other incoming node base should match the B base.
+      // SecondInstOffset >= OffsetB && FirstInstOffset > 0?
+      // SecondInstOffset <= OffsetB && FirstInstOffset < 0?
+      // Is non-equal if above are true.
+      APInt FirstInstOffset(Q.DL.getIndexTypeSizeInBits(FirstInst->getType()),
+                            0);
+      FirstInst = FirstInst->stripAndAccumulateConstantOffsets(
+          Q.DL, FirstInstOffset, /* AllowNonInbounds */ true);
+      APInt SecondInstOffset(Q.DL.getIndexTypeSizeInBits(SecondInst->getType()),
+                             0);
+      SecondInst = SecondInst->stripAndAccumulateConstantOffsets(
+          Q.DL, SecondInstOffset, /* AllowNonInbounds */ true);
+      APInt OffsetB(Q.DL.getIndexTypeSizeInBits(B->getType()), 0);
+      B = B->stripAndAccumulateConstantOffsets(Q.DL, OffsetB,
+                                               /* AllowNonInbounds */ true);
+      if (SecondInst == B &&
+          ((SecondInstOffset.sge(OffsetB) && FirstInstOffset.sgt(0)) ||
```

Comment:
Doesn't there need to be a check that `SecondInst` has a constant offset? Otherwise seems to me you could end up saying that: `(gep A_ptr, C_index) != (gep B_ptr, V_index)`

Reply:
Thanks for the review. 
Since we have checked at the start that both A and B are of pointer types, we can always get a const offset(zero or non-zero). So we can have scenarios where we can then handle both test recusrsiveGEP_withPtr_phisub1 where SecondInst is a ptr with zero offset, and recursiveGEP_withPtr_phisub3 where SecondInst is a ptr of type GEP.